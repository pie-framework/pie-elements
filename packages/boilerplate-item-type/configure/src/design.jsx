import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import cloneDeep from 'lodash/cloneDeep';
import { InputContainer, settings, layout } from '@pie-lib/pie-toolbox/config-ui';
import {EditableHtml} from '@pie-lib/pie-toolbox/editable-html';
import {test, callTest} from 'pie-utils';

callTest('configure');

const { Panel, toggle } = settings;

export class Design extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    uploadSoundSupport: PropTypes.object,
  };

  static defaultProps = {};

  onPromptChanged = (prompt) => {
    const { onModelChanged, model } = this.props;
    const update = cloneDeep(model);

    update.prompt = prompt;
    onModelChanged(update);
  };

  render() {
    const { classes, configuration, imageSupport, model, onConfigurationChanged, onModelChanged, uploadSoundSupport } =
      this.props;
    const {
      contentDimensions = {},
      prompt = {},
      settingsPanelDisabled,
    } = configuration || {};
    const { promptEnabled } = model || {};

    const panelSettings = {};

    const panelProperties = {
      promptEnabled: prompt.settings && toggle(prompt.label),
    };

    return (
      <layout.ConfigLayout
        dimensions={contentDimensions}
        hideSettings={settingsPanelDisabled}
        settings={
          <Panel
            model={model}
            configuration={configuration}
            onChangeModel={(model) => onModelChanged(model)}
            onChangeConfiguration={onConfigurationChanged}
            groups={{
              Settings: panelSettings,
              Properties: panelProperties,
            }}
          />
        }
      >
        {test}
        {promptEnabled && (
          <InputContainer label={prompt.label || ''} className={classes.promptHolder}>
            <EditableHtml
              className={classes.prompt}
              markup={model.prompt}
              onChange={this.onPromptChanged}
              imageSupport={imageSupport}
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
            />
          </InputContainer>
        )}
      </layout.ConfigLayout>
    );
  }
}

export default withStyles((theme) => ({
  promptHolder: {
    width: '100%',
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
  }
}))(Design);
