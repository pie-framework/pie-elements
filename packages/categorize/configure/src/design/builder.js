import cloneDeep from 'lodash/cloneDeep';

export const buildCategories = (categories, choices, correctResponse) => {
  const clonedCategories = cloneDeep(categories);

  return clonedCategories.map(category => {
    const cr = correctResponse.find(cr => cr.category === category.id);

    if (cr) {
      category.choices = (cr.choices || []).map(choiceId => {
        const choice = choices.find(h => h.id === choiceId);
        if (choice) {
          return Object.assign({}, { id: choice.id, content: choice.content });
        }
      });
    }

    return category;
  });
};

const getChoices = (cat, choices, index) => {
  if (!cat.alternateResponses || cat.alternateResponses.length === 0) {
    return [];
  }

  return (cat.alternateResponses[index] || []).map(alt => {
    return choices.find(ch => ch.id === alt);
  });
};

export const buildAlternateResponses = (categories, choices, correctResponse) => {
  const mostAlternates = correctResponse.reduce((mostAlt, cat) => {
    if (cat.alternateResponses && cat.alternateResponses.length >= mostAlt) {
      return cat.alternateResponses.length;
    }

    return mostAlt;
  }, 0);
  const alternatesArray = (new Array(mostAlternates)).fill(0);

  return alternatesArray.map((val, index) => {
    return correctResponse.map((cat, catIndex) => {
      const currentCategory = categories[catIndex];

      return {
        id: currentCategory.id,
        label: currentCategory.label,
        choices: getChoices(cat, choices, index)
      };
    });
  });
};
