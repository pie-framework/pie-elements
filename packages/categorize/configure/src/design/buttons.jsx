import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Button from '@material-ui/core/Button';
import MuiDivider from '@material-ui/core/Divider';

export class RawAddButton extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    label: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    label: 'Add'
  };

  render() {
    const { classes, className, label, onClick, disabled } = this.props;
    return (
      <Button
        onClick={onClick}
        disabled={disabled}
        className={classNames(classes.addButton, className)}
        size="small"
        variant="contained"
        color="primary"
      >
        {label}
      </Button>
    );
  }
}
const styles = theme => ({
  addButton: {
    height: theme.spacing.unit * 4
  }
});

const AddButton = withStyles(styles)(RawAddButton);

const DeleteButton = withStyles(theme => ({
  deleteButton: {
    margin: 0,
    padding: 0
  }
}))(({ classes, label, onClick }) => (
  <Button
    className={classes.deleteButton}
    onClick={onClick}
    size="small"
    color="primary"
  >
    {label}
  </Button>
));

const Divider = withStyles(theme => ({
  divider: {
    marginTop: 2 * theme.spacing.unit,
    marginBottom: 2 * theme.spacing.unit
  }
}))(({ classes }) => <MuiDivider className={classes.divider} />);

export { AddButton, DeleteButton, Divider };
