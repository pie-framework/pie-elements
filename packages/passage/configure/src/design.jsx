import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { InputContainer, settings, layout } from '@pie-lib/pie-toolbox/config-ui';
import { EditableHtml, ALL_PLUGINS } from '@pie-lib/pie-toolbox/editable-html';

const { Panel, toggle, dropdown } = settings;

export class Main extends React.Component {
  static propTypes = {
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func,
    model: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
    imageSupport: PropTypes.object.isRequired,
    uploadSoundSupport: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = { setDimensions: true };
  }

  handleChange = (fieldName, value, index = 0) => {
    const { model, onModelChanged } = this.props;

    if (!model || !onModelChanged || !model.passages || index < 0 || index >= model.passages.length) {
      return;
    }

    const updatedPassages = [...model.passages];
    updatedPassages[index] = { ...updatedPassages[index], [fieldName]: value };

    onModelChanged({ ...model, passages: updatedPassages });
  };

  render() {
    const { model, classes, configuration, imageSupport, onConfigurationChanged, onModelChanged, uploadSoundSupport } =
      this.props;
    const {
      settingsPanelDisabled,
      language = {},
      languageChoices = {},
      maxImageWidth = {},
      maxImageHeight = {},
      mathMlOptions = {},
      baseInputConfiguration = {},
      teacherInstructions = {},
      title = {},
      subtitle = {},
      text = {},
      author = {},
    } = configuration || {};
    const {
      errors = {},
      extraCSSRules,
      passages = [],
      teacherInstructionsEnabled,
      titleEnabled,
      subtitleEnabled,
      authorEnabled,
      textEnabled,
    } = model || {};

    const { passages: passagesErrors } = errors || {};
    const {
      teacherInstructions: teacherInstructionsError,
      title: titleError,
      subtitle: subtitleError,
      author: authorError,
      text: textError,
    } = (passagesErrors && passagesErrors[0]) || {}; // only the first passage errors are needed for now

    const defaultImageMaxWidth = maxImageWidth && maxImageWidth.prompt;
    const defaultImageMaxHeight = maxImageHeight && maxImageHeight.prompt;

    const panelSettings = {
      titleEnabled: title && title.settings && toggle(title.label),
      subtitleEnabled: subtitle && subtitle.settings && toggle(subtitle.label),
      authorEnabled: author && author.settings && toggle(author.label),
      textEnabled: text && text.settings && toggle(text.label),
    };

    const panelProperties = {
      teacherInstructionsEnabled:
        teacherInstructions && teacherInstructions.settings && toggle(teacherInstructions.label),
      'language.enabled': language && language.settings && toggle(language.label, true),
      language:
        language && language.settings && language.enabled && dropdown(languageChoices.label, languageChoices.options),
    };

    const getPluginProps = (customConfiguration) => {
      return {
        ...baseInputConfiguration,
        ...customConfiguration,
      };
    };

    return (
      <layout.ConfigLayout
        extraCSSRules={extraCSSRules}
        hideSettings={settingsPanelDisabled}
        settings={
          <Panel
            model={model}
            configuration={configuration}
            onChangeModel={(model) => onModelChanged(model)}
            onChangeConfiguration={(config) => onConfigurationChanged(config)}
            groups={{
              Settings: panelSettings,
              Properties: panelProperties,
            }}
          />
        }
      >
        {teacherInstructionsEnabled && (
          <InputContainer label={teacherInstructions.label} className={classes.inputContainer}>
            <EditableHtml
              activePlugins={ALL_PLUGINS}
              markup={passages[0].teacherInstructions || ''}
              onChange={(value) => this.handleChange('teacherInstructions', value)}
              nonEmpty={false}
              error={teacherInstructionsError}
              maxImageWidth={(maxImageWidth && maxImageWidth.teacherInstructions) || defaultImageMaxWidth}
              maxImageHeight={(maxImageHeight && maxImageHeight.teacherInstructions) || defaultImageMaxHeight}
              mathMlOptions={mathMlOptions}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
              pluginProps={getPluginProps(teacherInstructions?.inputConfiguration)}
            />
            {teacherInstructionsError && <div className={classes.errorText}>{teacherInstructionsError}</div>}
          </InputContainer>
        )}

        {titleEnabled && (
          <InputContainer label={title.label} className={classes.inputContainer}>
            <EditableHtml
              activePlugins={ALL_PLUGINS}
              markup={passages[0].title || ''}
              onChange={(value) => this.handleChange('title', value)}
              nonEmpty={false}
              error={titleError}
              mathMlOptions={mathMlOptions}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
              pluginProps={getPluginProps(title?.inputConfiguration)}
            />
            {titleError && <div className={classes.errorText}>{titleError}</div>}
          </InputContainer>
        )}

        {subtitleEnabled && (
          <InputContainer label={subtitle.label} className={classes.inputContainer}>
            <EditableHtml
              activePlugins={ALL_PLUGINS}
              markup={passages[0].subtitle || ''}
              onChange={(value) => this.handleChange('subtitle', value)}
              nonEmpty={false}
              error={subtitleError}
              mathMlOptions={mathMlOptions}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
              pluginProps={getPluginProps(subtitle?.inputConfiguration)}
            />
            {subtitleError && <div className={classes.errorText}>{subtitleError}</div>}
          </InputContainer>
        )}

        {authorEnabled && (
          <InputContainer label={author.label} className={classes.inputContainer}>
            <EditableHtml
              activePlugins={ALL_PLUGINS}
              markup={passages[0].author || ''}
              onChange={(value) => this.handleChange('author', value)}
              nonEmpty={false}
              error={authorError}
              mathMlOptions={mathMlOptions}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
              pluginProps={getPluginProps(author?.inputConfiguration)}
            />
            {authorError && <div className={classes.errorText}>{authorError}</div>}
          </InputContainer>
        )}

        {textEnabled && (
          <InputContainer label={text.label} className={classes.inputContainer}>
            <EditableHtml
              activePlugins={ALL_PLUGINS}
              markup={passages[0].text || ''}
              onChange={(value) => this.handleChange('text', value)}
              imageSupport={imageSupport}
              uploadSoundSupport={uploadSoundSupport}
              nonEmpty={false}
              error={textError}
              maxImageWidth={(maxImageWidth && maxImageWidth.text) || defaultImageMaxWidth}
              maxImageHeight={(maxImageHeight && maxImageHeight.text) || defaultImageMaxHeight}
              mathMlOptions={mathMlOptions}
              languageCharactersProps={[{ language: 'spanish' }, { language: 'special' }]}
              pluginProps={getPluginProps(text?.inputConfiguration)}
            />
            {textError && <div className={classes.errorText}>{textError}</div>}
          </InputContainer>
        )}
      </layout.ConfigLayout>
    );
  }
}
export default withStyles((theme) => ({
  inputContainer: {
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
    width: '100%',
  },
  errorText: {
    fontSize: theme.typography.fontSize - 2,
    color: theme.palette.error.main,
    paddingTop: theme.spacing.unit,
  },
}))(Main);
