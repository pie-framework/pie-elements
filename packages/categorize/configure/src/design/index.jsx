import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Typography from '@material-ui/core/Typography';
import {
  FeedbackConfig,
  InputContainer,
  layout,
  settings,
} from '@pie-lib/config-ui';
import {
  countInAnswer,
  ensureNoExtraChoicesInAnswer
} from '@pie-lib/categorize';
import EditableHtml from '@pie-lib/editable-html';
import { uid, withDragContext } from '@pie-lib/drag';
import debug from 'debug';

import Categories from './categories';
import AlternateResponses from './categories/alternateResponses';
import Choices from './choices';
import { Divider } from './buttons';
import {
  buildAlternateResponses,
  buildCategories
} from './builder';
import Header from './header';

const { Panel, toggle, radio } = settings;
const { Provider: IdProvider } = uid;

export class Design extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    uid: PropTypes.string,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    })
  };

  constructor(props) {
    super(props);
    this.uid = props.uid || uid.generateId();
  }

  updateModel = props => {
    const { model, onChange } = this.props;

    const updatedModel = {
      ...model,
      ...props
    };

    //Ensure that there are no extra choices in correctResponse, if the user has decided that only one choice may be used.
    updatedModel.correctResponse = ensureNoExtraChoicesInAnswer(
      updatedModel.correctResponse || [],
      updatedModel.choices
    );

    //clean categories
    updatedModel.categories = updatedModel.categories.map(c => ({
      id: c.id,
      label: c.label
    }));

    updatedModel.choices = updatedModel.choices.map(h => ({
      id: h.id,
      content: h.content,
      categoryCount: h.categoryCount
    }));

    onChange(updatedModel);
  };

  changeRationale = rationale => {
    const { model, onChange } = this.props;

    onChange({
      ...model,
      rationale
    });
  };

  changeTeacherInstructions = teacherInstructions => {
    const { model, onChange } = this.props;

    onChange({
      ...model,
      teacherInstructions
    });
  };

  changeFeedback = feedback => {
    this.updateModel({ feedback });
  };

  onAddAlternateResponse = () => {
    const { model: { correctResponse } } = this.props;

    this.updateModel({
      correctResponse: (correctResponse || []).map(cr => ({
        ...cr,
        alternateResponses: [
          ...(cr.alternateResponses || []),
          []
        ]
      }))
    });
  };

  onPromptChanged = prompt => this.updateModel({ prompt });

  onRemoveAlternateResponse = (index) => {
    const { model: { correctResponse } } = this.props;

    this.updateModel({
      correctResponse: (correctResponse || []).map((cr) => ({
        ...cr,
        alternateResponses: (cr.alternateResponses || []).filter((alt, altIndex) => altIndex !== index)
      }))
    });
  };

  countChoiceInCorrectResponse = choice => {
    const { model } = this.props;
    const out = countInAnswer(choice.id, model.correctResponse);
    return out;
  };

  render() {
    const { classes, className, model, imageSupport, configuration, onChange, onConfigurationChanged } = this.props;
    const {
      partialScoring = {},
      lockChoiceOrder = {},
      teacherInstructions = {},
      studentInstructions = {},
      rationale = {},
      scoringType = {},
      feedback = {},
      prompt = {}
    } = configuration || {};

    const config = model.config || {};
    config.choices = config.choices || { label: '', columns: 2 };

    const categories = buildCategories(
      model.categories || [],
      model.choices || [],
      model.correctResponse || []
    );

    const alternateResponses = buildAlternateResponses(
      model.categories || [],
      model.choices || [],
      model.correctResponse || []
    );

    const choices = model.choices.map(c => {
      c.correctResponseCount = this.countChoiceInCorrectResponse(c);
      return c;
    });

    return (
      <IdProvider value={this.uid}>
        <layout.ConfigLayout
          settings={
            <Panel
              model={model}
              onChangeModel={onChange}
              configuration={configuration}
              onChangeConfiguration={onConfigurationChanged}
              groups={{
                'Settings': {
                  partialScoring: partialScoring.settings &&
                    toggle(partialScoring.label),
                  lockChoiceOrder: lockChoiceOrder.settings &&
                  toggle(lockChoiceOrder.label),
                  'feedback.enabled': feedback.settings && toggle(feedback.label, true),
                },
                'Properties': {
                  'teacherInstructions.enabled': teacherInstructions.settings &&
                    toggle(teacherInstructions.label, true),
                  'studentInstructions.enabled': studentInstructions.settings &&
                    toggle(studentInstructions.label, true),
                  'rationale.enabled': rationale.settings &&
                    toggle(rationale.label, true),
                  scoringType: scoringType.settings &&
                    radio(scoringType.label, ['auto', 'rubric']),
                },
              }}
            />
          }
        >
          <div className={classNames(classes.design, className)}>

            {prompt.settings && (
              <InputContainer
                label={prompt.label}
                className={classes.promptHolder}
              >
                <EditableHtml
                  className={classes.prompt}
                  markup={model.prompt}
                  onChange={this.onPromptChanged}
                  imageSupport={imageSupport}
                  nonEmpty={!prompt.settings}
                  disableUnderline
                />
              </InputContainer>
            )}

            {teacherInstructions.enabled && (
              <InputContainer label={teacherInstructions.label} className={classes.inputHolder}>
                <EditableHtml
                  className={classes.input}
                  markup={model.teacherInstructions || ''}
                  onChange={this.changeTeacherInstructions}
                  imageSupport={imageSupport}
                  nonEmpty={false}
                />
              </InputContainer>
            )}

            {rationale.enabled && (
              <InputContainer label={rationale.label} className={classes.inputHolder}>
                <EditableHtml
                  className={classes.input}
                  markup={model.rationale || ''}
                  onChange={this.changeRationale}
                  imageSupport={imageSupport}
                  nonEmpty={false}
                />
              </InputContainer>
            )}

            <Categories
              imageSupport={imageSupport}
              model={model}
              categories={categories}
              onModelChanged={this.updateModel}
            />

            <Header
              className={classes.alternatesHeader}
              label="Alternate Responses"
              buttonLabel="ADD AN ALTERNATE RESPONSE"
              onAdd={this.onAddAlternateResponse}
            />

            {
              alternateResponses.map((categoriesList, index) => {
                return (
                  <React.Fragment key={index}>
                    <Header
                      className={classes.alternatesHeader}
                      label="Alternate Response"
                      buttonLabel="REMOVE ALTERNATE RESPONSE"
                      onAdd={() => this.onRemoveAlternateResponse(index)}
                    />

                    <AlternateResponses
                      altIndex={index}
                      imageSupport={imageSupport}
                      model={model}
                      categories={categoriesList}
                      onModelChanged={this.updateModel}
                    />

                  </React.Fragment>
                )
              })
            }

            <Divider />
            <Choices
              imageSupport={imageSupport}
              choices={choices}
              model={model}
              onModelChanged={this.updateModel}
            />

            {
              feedback.enabled && (
                <FeedbackConfig
                  feedback={model.feedback}
                  onChange={this.changeFeedback}
                />
              )
            }
          </div>
        </layout.ConfigLayout>
      </IdProvider>
    );
  }
}

const styles = theme => ({
  alternatesHeader: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  text: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  design: {
    paddingTop: theme.spacing.unit,
    paddingBottom: theme.spacing.unit
  },
  inputHolder: {
    width: '100%',
    paddingBottom: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2
  },
  input: {
    paddingTop: theme.spacing.unit * 2,
    width: '100%',
    maxWidth: '600px'
  },
  prompt: {
    paddingTop: theme.spacing.unit * 2,
    width: '100%'
  },
  promptHolder: {
    width: '100%',
    paddingBottom: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2
  },
  title: {
    marginBottom: '30px'
  }
});

export default withDragContext(withStyles(styles)(Design));
