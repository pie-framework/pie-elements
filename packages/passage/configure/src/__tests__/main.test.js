import React from 'react';
import { shallow } from 'enzyme';
import { Main } from '../design';

jest.mock('@pie-lib/pie-toolbox/config-ui', () => ({
  layout: {
    ConfigLayout: (props) => <div className="mockConfigLayout">{props.children}</div>,
  },
  settings: {
    Panel: (props) => <div className="mockPanel" onClick={props.onChange} />,
    toggle: (label, defaultVal = false) => true,
    dropdown: jest.fn(),
  },
}));

jest.mock('../common', () => ({
  ConfimationDialog: (props) => <div className="mockDialog" {...props} />,
  PassageButton: (props) => (
    <button className="mockPassageButton" onClick={props.onClick}>
      {props.label}
    </button>
  ),
}));

jest.mock('../passage', () => (props) => <div className="mockPassage" {...props} />);

describe('Main Component - Unit Method Tests', () => {
  let wrapper, instance, onModelChanged;

  const basePassage = {
    teacherInstructions: '',
    title: '',
    subtitle: '',
    author: '',
    text: '',
  };

  const baseModel = {
    passages: [basePassage],
    language: 'en',
  };

  const configuration = {
    additionalPassage: { label: 'Additional Passage', settings: true, enabled: true },
    teacherInstructions: { label: 'Teacher Instructions', settings: true },
    title: { label: 'Title', settings: true },
    subtitle: { label: 'Subtitle', settings: true },
    author: { label: 'Author', settings: true },
    text: { label: 'Text', settings: true },
  };

  const props = {
    model: baseModel,
    configuration,
    imageSupport: {},
    uploadSoundSupport: {},
    classes: { additionalPassageHeading: 'heading' },
    onModelChanged: jest.fn(),
    onConfigurationChanged: jest.fn(),
  };

  beforeEach(() => {
    onModelChanged = jest.fn();
    wrapper = shallow(<Main {...props} onModelChanged={onModelChanged} />);
    instance = wrapper.instance();
  });

  it('getInnerText removes HTML tags', () => {
    expect(instance.getInnerText('<p>Test <strong>123</strong></p>')).toBe('Test 123');
    expect(instance.getInnerText('')).toBe('');
    expect(instance.getInnerText(null)).toBe('');
    expect(instance.getInnerText('<img src="x" />')).toBe('');
  });

  it('addAdditionalPassage adds an empty passage and calls onModelChanged', () => {
    instance.addAdditionalPassage();

    expect(onModelChanged).toHaveBeenCalledWith({
      ...baseModel,
      passages: [...baseModel.passages, { teacherInstructions: '', title: '', subtitle: '', author: '', text: '' }],
    });
  });

  it('removeAdditionalPassage calls onDelete directly for empty passage', () => {
    const modelWithTwo = {
      ...baseModel,
      passages: [basePassage, basePassage],
    };

    wrapper.setProps({ model: modelWithTwo });
    instance = wrapper.instance();

    const onDeleteSpy = jest.spyOn(instance, 'onDelete');
    instance.removeAdditionalPassage(1);

    expect(onDeleteSpy).toHaveBeenCalledWith(modelWithTwo, 1, onModelChanged);
  });

  it('removeAdditionalPassage sets dialog state if passage has content', () => {
    const modelWithContent = {
      ...baseModel,
      passages: [basePassage, { ...basePassage, teacherInstructions: '<p>Filled</p>' }],
    };

    wrapper.setProps({ model: modelWithContent });
    instance.removeAdditionalPassage(1);

    expect(wrapper.state()).toEqual({
      showConfirmationDialog: true,
      indexToRemove: 1,
    });
  });

  it('onDelete removes passage at index and resets confirmation state', () => {
    const modelWithTwo = {
      ...baseModel,
      passages: [{ teacherInstructions: 'Keep' }, { teacherInstructions: 'Remove' }],
    };

    wrapper.setProps({ model: modelWithTwo });
    instance.onDelete(modelWithTwo, 1, onModelChanged);

    expect(onModelChanged).toHaveBeenCalledWith({
      ...modelWithTwo,
      passages: [{ teacherInstructions: 'Keep' }],
    });

    expect(wrapper.state()).toEqual({
      showConfirmationDialog: false,
      indexToRemove: -1,
    });
  });
});
