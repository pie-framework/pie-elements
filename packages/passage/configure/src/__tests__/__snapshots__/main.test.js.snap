// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Main Component Match Snapshot 1`] = `
<ConfigLayout
  hideSettings={false}
  settings={
    <Panel
      configuration={
        Object {
          "author": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": true,
              },
              "image": Object {
                "disabled": true,
              },
              "video": Object {
                "disabled": true,
              },
            },
            "label": "Author",
            "required": false,
            "settings": true,
          },
          "baseInputConfiguration": Object {
            "audio": Object {
              "disabled": false,
            },
            "image": Object {
              "disabled": false,
            },
            "video": Object {
              "disabled": false,
            },
          },
          "language": Object {
            "enabled": true,
            "label": "Specify Language",
            "settings": true,
          },
          "languageChoices": Object {
            "label": "Language Choices",
            "options": Array [],
          },
          "mathMlOptions": Object {
            "mmlEditing": false,
            "mmlOutput": false,
          },
          "maxImageHeight": Object {
            "teacherInstructions": 300,
            "text": 300,
          },
          "maxImageWidth": Object {
            "teacherInstructions": 300,
            "text": 300,
          },
          "settingsPanelDisabled": false,
          "subtitle": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": true,
              },
              "image": Object {
                "disabled": true,
              },
              "video": Object {
                "disabled": true,
              },
            },
            "label": "Subtitle",
            "required": false,
            "settings": true,
          },
          "teacherInstructions": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
            "label": "Teacher Instructions",
            "required": false,
            "settings": true,
          },
          "text": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
            "label": "Text",
            "required": false,
            "settings": true,
          },
          "title": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": true,
              },
              "image": Object {
                "disabled": true,
              },
              "video": Object {
                "disabled": true,
              },
            },
            "label": "Title",
            "required": false,
            "settings": true,
          },
        }
      }
      groups={
        Object {
          "Properties": Object {
            "language": undefined,
            "language.enabled": undefined,
            "teacherInstructionsEnabled": undefined,
          },
          "Settings": Object {
            "authorEnabled": undefined,
            "subtitleEnabled": undefined,
            "textEnabled": undefined,
            "titleEnabled": undefined,
          },
        }
      }
      model={
        Object {
          "authorEnabled": true,
          "passages": Array [
            Object {
              "author": "",
              "subtitle": "",
              "teacherInstructions": "",
              "text": "",
              "title": "",
            },
          ],
          "subtitleEnabled": true,
          "teacherInstructionsEnabled": false,
          "textEnabled": true,
          "titleEnabled": true,
        }
      }
      onChangeConfiguration={[Function]}
      onChangeModel={[Function]}
    />
  }
>
  <InputContainer
    className="mockInputContainer"
    label="Title"
  >
    <EditableHtml
      allowValidation={false}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
    />
  </InputContainer>
  <InputContainer
    className="mockInputContainer"
    label="Subtitle"
  >
    <EditableHtml
      allowValidation={false}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
    />
  </InputContainer>
  <InputContainer
    className="mockInputContainer"
    label="Author"
  >
    <EditableHtml
      allowValidation={false}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": true,
          },
        }
      }
    />
  </InputContainer>
  <InputContainer
    className="mockInputContainer"
    label="Text"
  >
    <EditableHtml
      activePlugins={
        Array [
          "bold",
          "html",
          "italic",
          "underline",
          "strikethrough",
          "bulleted-list",
          "numbered-list",
          "image",
          "math",
          "languageCharacters",
          "table",
          "video",
          "audio",
          "responseArea",
        ]
      }
      allowValidation={false}
      imageSupport={Object {}}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      maxImageHeight={300}
      maxImageWidth={300}
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": false,
          },
          "image": Object {
            "disabled": false,
          },
          "video": Object {
            "disabled": false,
          },
        }
      }
      uploadSoundSupport={Object {}}
    />
  </InputContainer>
</ConfigLayout>
`;
